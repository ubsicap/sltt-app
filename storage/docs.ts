import { createHash } from 'crypto'
import { basename, join, parse, sep } from 'path'
import { existsSync } from 'fs'
import { readFile, writeFile, readdir, unlink, appendFile, stat, open } from 'fs/promises'
import { ensureDir, ensureFile, readJson, writeJson, read, close } from 'fs-extra'
import { sortBy } from 'lodash'
import { promisify } from 'util'
import { ListDocsArgs, ListDocsResponse, RemoteSeqDoc, RetrieveDocArgs, RetrieveDocResponse, StoreDocArgs, StoreDocResponse, SyncDocsArgs, SyncDocsResponse, SyncRemoteDocsArgs } from './docs.d'
import { readJsonCatchMissing } from './utils'


const readAsync = promisify(read)
const closeAsync = promisify(close)

async function readLastBytes(filePath: string, byteCount: number): Promise<Buffer> {
    // Open the file in read mode
    const fileHandle = await open(filePath, 'r')
    try {
        // Get the size of the file
        const fileStats = await stat(filePath)
        const fileSize = fileStats.size

        // Calculate the position to start reading from
        const startPosition = Math.max(0, fileSize - byteCount)

        // Create a buffer to hold the bytes
        const buffer = Buffer.alloc(byteCount)

        // Read the bytes from the file
        await readAsync(fileHandle.fd, buffer, 0, byteCount, startPosition)

        return buffer
    } finally {
        // Close the file
        await closeAsync(fileHandle.fd)
    }
}

const composeFilenameSafeDate = (modDate: string): string => {
    let dateStr = modDate // 2024/06/17 09:49:07.997Z
    // Replace slashes, spaces, and colons with underscores to make it filename-safe
    dateStr = dateStr.replace(/\//g, '-') // Replace slashes with hyphens
    dateStr = dateStr.replace(/ /g, '_') // Replace spaces with underscores
    dateStr = dateStr.replace(/:/g, '-') // Replace colons with hyphens
    // Handle milliseconds and 'Z' - replace '.' with '-' and remove 'Z'
    dateStr = dateStr.replace(/\./g, '-').replace(/Z$/, '')
    return dateStr // 2024-06-17_09-49-07-997
}

const decomponseFilenameSafeDate = (filenameSafeDate: string): string => {
    let [dateStr, timeStr] = filenameSafeDate.split('_') // 2024-06-17, 09-49-07-997 <-- 2024-06-17_09-49-07-997
    // Replace hyphens with slashes, spaces, and colons to make it filename-safe
    dateStr = dateStr.replace(/-/g, '/') // Replace hyphens with slashes
    timeStr = timeStr.replace(/-/g, ':') // Replace hyphens with colons
    // Handle milliseconds and 'Z' - replace '-' with '.' and add 'Z'
    timeStr = timeStr.replace(/-/g, '.').concat('Z')
    return `${dateStr} ${timeStr}` // 2024/06/17 09:49:07.997Z
}

const composeFilenameSafeId = (_id: string): string => {
    // GIVEN _id in format like plan_240617_094907/stg_240617_094910/tsk_240617_094912
    // Replace slashes with hyphens
    if (!_id) return 'no-id'
    const filenameSafeId1 = _id.replace(/\//g, '-') // plan_240617_094907-stg_240617_094910-tsk_240617_094912
    const fileNameExtra = composeFilename('9999/99/99 99:99:99.999Z', '', '999@99999.999', '999@99999.999', 999999999)
    const fileNameTrial1 = fileNameExtra.replace('no-id', filenameSafeId1)
    if (fileNameTrial1.length >= 255) {
        // if filename is too long, shorten each inner date_time to just the time component, but keep first and last parts of _id
        const abbreviatedId = _id.split('/').map((s, i, array) => (i > 0 && i < array.length - 1) ? s.split('_').slice(-1) : s).join('-')
        return abbreviatedId // plan_240617_094907-094910-tsk_240617_094912 <-- plan_240617_094907/stg_240617_094910/tsk_240617_094912
    } else {
        return filenameSafeId1 // plan_240617_094907-stg_240617_094910-tsk_240617_094912 <-- plan_240617_094907/stg_240617_094910/tsk_240617_094912
    }
}

const LOCAL_DOC_PREFIX = 'local-doc' // 9 characters...same as remote seq

const composeFilename = (modDate: string, _id: string, creator: string, modBy: string, remoteSeq: number): string => {
    const filenameSafeModDate = composeFilenameSafeDate(modDate)
    const filenameSafeId = composeFilenameSafeId(_id)
    const filenameSafeCreator = composeFilenameSafeEmail(creator)
    const filenameSafeModBy = modBy && composeFilenameSafeEmail(modBy) || 'no-mod-by'
    // make slot for remote and local nine characters so abbreviation logic is applied to both
    // this can make determining whether local has become remote based on the filename alone) 
    const filenameRemoteSeq = Number.isInteger(remoteSeq) ? `${`${remoteSeq}`.padStart(9, '0')}` : LOCAL_DOC_PREFIX
    const filename = `${filenameRemoteSeq}__${filenameSafeModDate}__${filenameSafeId}__${filenameSafeCreator}__${filenameSafeModBy}.sltt-doc`
    return filename
}

const decomposeRemoteSeq = (paddedRemoteSeq: string): string =>
    paddedRemoteSeq === 'local' ? '' : paddedRemoteSeq.replace(/^0+/, '')

const createMd5Hash = (s: string): string => createHash('md5').update(s).digest('hex').toString()
const createEmailHash = (email: string): string => createMd5Hash(email).substring(0, 16) // api uses this

const composeFilenameSafeEmail = (email: string): string => {
    return createEmailHash(email).substring(0, 8) // 8 characters will probably avoid collision within team 
}

const buildDocFolder = (docsFolder: string, project: string, isFromRemote: boolean): string => {
    const DOCS_FROM_REMOTE_PATH = 'remote'
    const DOCS_FROM_LOCAL_PATH = 'local'
    const fullFromPath = isFromRemote ? DOCS_FROM_REMOTE_PATH : DOCS_FROM_LOCAL_PATH
    if (!docsFolder) {
        throw Error('docsPath not set')
    }
    return join(docsFolder, basename(project), fullFromPath)
}

const parseFilename = (filename: string): { projectPath: string, normalizedFilename: string, remoteSeq: string, filenameModDate: string, filenameId: string, filenameCreator: string, filenameModBy: string } => {
    const projectPath = filename.split(sep).slice(-3, -1).join('/')  // {project}/{local|remote}
    const normalizedFilename = basename(filename) // prevent path traversal
    const filenameWithoutExt = parse(normalizedFilename).name
    const [remoteSeq, filenameModDate, filenameId, filenameCreator, filenameModBy] = filenameWithoutExt.split('__')
    return { projectPath, normalizedFilename, remoteSeq, filenameModDate, filenameId, filenameCreator, filenameModBy }
}

type IDBObject = { _id: string, modDate: string, creator: string, modBy?: string }

export const handleStoreDocV0 = async (docsFolder: string, { clientId, project, doc, remoteSeq }: StoreDocArgs<IDBObject>):
    Promise<StoreDocResponse> => {
    if (remoteSeq > 999999999) {
        throw Error(`remoteSeq is too large: ${remoteSeq}`)
    }
    const isFromRemote = !!remoteSeq
    const isLocalDoc = !isFromRemote
    const fullFromPath = buildDocFolder(docsFolder, project, isFromRemote)
    const { _id, modDate, creator, modBy } = doc as { _id: string, modDate: string, creator: string, modBy: string }
    if (!_id || !modDate) {
        throw Error(`_id and modDate properties are required in doc: ${JSON.stringify(doc)}`)
    }
    if (isLocalDoc && !modBy) {
        throw Error(`modBy property is required in local doc: ${JSON.stringify(doc)}`)
    }
    const filename = composeFilename(modDate, _id, creator, modBy, remoteSeq)
    if (filename.length > 255) {
        throw Error(`attempted filename is too long: ${filename}`)
    }
    await ensureDir(fullFromPath)
    let finalFilename = filename
    if (!remoteSeq) {
        // see if _id has already been stored locally with a later modDate
        // if so, add `-lost` to the filename
        // TODO: cache listDocs and maintain it in memory
        try {
            const localFilenames = await listDocs(docsFolder, {
                project, isFromRemote: false,
                fnFilter: (storedFilename) => storedFilename.split('__')[2] === filename.split('__')[2]
            })
            if (localFilenames.includes(filename)) {
                // filename already exists locally, so don't overwrite it
                return { ...parseFilename(filename), freshlyWritten: false }
            }
            // sort localFilenames and get modDate from last one
            const mostRecentLocalFilename = [...localFilenames, filename].sort().pop()
            if (mostRecentLocalFilename !== filename) {
                const lostFilename = `${filename}-lost`
                finalFilename = lostFilename
            }
            const fullPath = join(fullFromPath, finalFilename)
            return await writeDoc(fullPath, doc)
        } catch (error) {
            console.error('An error occurred:', error.message)
        }
    }
    const fullPath = join(fullFromPath, finalFilename)
    const response = writeDoc(fullPath, doc)
    if (remoteSeq) {  //Is from remote is true
        // see if a corresponding local-doc file exists
        const localFolder = buildDocFolder(docsFolder, project, false)
        const localFilename = composeFilename(modDate, _id, creator, modBy, Number.NaN)
        const localPath = join(localFolder, localFilename)
        try {
            // Check if the file exists in the local folder
            // await fs.access(localPath)
            // When the file exists, delete it
            await unlink(localPath)
            console.log(`Successfully deleted local file: ${localPath}`)
        } catch (error) {
            if (error.code === 'ENOENT') {
                console.log(`Local file does not exist: ${localPath}`)
            } else {
                console.error(`Failed to delete local file: ${localPath}`, error)
            }
        }
    }
    return response
}

export const handleSyncRemoteDocs = async (
    docsFolder: string, { clientId, project, seqDocs }: SyncRemoteDocsArgs<IDBObject>)
    : Promise<SyncDocsResponse<IDBObject>> => {
    
    const fullFromPath = buildDocFolder(docsFolder, project, true)
    await ensureDir(fullFromPath)
    
    const remoteSeqDocsFile = join(fullFromPath, `remote.sltt-docs`)
    // read the last `000000000` characters from the file to get the last remoteSeq
    let nextRemoteSeq = 0
    try {
        const lastBytes = await readLastBytes(remoteSeqDocsFile, 9)
        nextRemoteSeq = Number(lastBytes.toString())
    } catch (error) {
        if (error.code === 'ENOENT') {
            // file doesn't exist, so it's the first sync
            console.log(`Remote file does not exist: ${remoteSeqDocsFile}`)
        } else {
            console.error(`Failed to read remote file: ${remoteSeqDocsFile}`, error)
        }
    }
    if (nextRemoteSeq === 0) {
        // append first nextRemoteSeq to the file
        await appendFile(remoteSeqDocsFile, '000000001')
        nextRemoteSeq = 1
    }
    // if any incoming seqDocs hava a remoteSeq greater or equal to nextRemoteSeq,
    // append them to the end of the file
    const newLines: string[] = []
    const sortedSeqDocs = sortBy(seqDocs, seqDoc => seqDoc.seq)
    for (const { doc, seq } of sortedSeqDocs) {
        if (seq >= nextRemoteSeq) {
            const newLine = `\t${Date.now()}\t${clientId}\t` + JSON.stringify(doc) + `\n${seq + 1}`
            newLines.push(newLine)
        }
    }
    if (newLines.length) {
        await appendFile(remoteSeqDocsFile, newLines.join(''))
    }
}

export const handleSyncLocalDocs = async (
    docsFolder: string, { clientId, project, docs, isRemote }: SyncDocsArgs<IDBObject>)
    : Promise<SyncDocsResponse<IDBObject>> => {
    
    const fullFromPath = buildDocFolder(docsFolder, project, isRemote)
    await ensureDir(fullFromPath)
}

const storeDocV1 = async (docsFolder: string, { clientId, project, doc, remoteSeq }: StoreDocArgs<IDBObject>): Promise<string> => {
    if (remoteSeq > 999999999) {
        throw Error(`remoteSeq is too large: ${remoteSeq}`)
    }
    const isFromRemote = !!remoteSeq
    const isLocalDoc = !isFromRemote
    const fullFromPath = buildDocFolder(docsFolder, project, isFromRemote)
    const { _id, modDate, modBy } = doc as { _id: string, modDate: string, creator: string, modBy: string }
    if (!_id || !modDate) {
        throw Error(`_id and modDate properties are required in doc: ${JSON.stringify(doc)}`)
    }
    if (isLocalDoc && !modBy) {
        throw Error(`modBy property is required in local doc: ${JSON.stringify(doc)}`)
    }
    const clientDocsPath = join(docsFolder, `${clientId}.sltt-docs`)
    try {
        await ensureFile(clientDocsPath)
        const status = isLocalDoc ? `  ` : '' // placeholder first character could be used for removing local docs that are in the remote list 
        // add terse utc timestamp (miliseconds after 2024-09-01) to each line
        // this will allow for sorting by time of creation
        const newLine = `${status}\t${Date.now()}\t` + JSON.stringify(doc) + '\n'
        await appendFile(fullFromPath, newLine)
        return newLine
    } catch (error) {
        console.error('An error occurred:', error.message)
        throw error
    }
}

export const handleListDocsV0 = async (docsFolder: string, { clientId, project, isFromRemote }: ListDocsArgs): Promise<ListDocsResponse> => {
    try {
        const filenames = await listDocs(docsFolder, { project, isFromRemote })
        if (!isFromRemote) {
            const localFilenames: string[] = []
            const remoteFilenames = await listDocs(docsFolder, { project, isFromRemote: true })
            const strippedRemoteFilenames = new Set(remoteFilenames.map((filename) => filename.slice(9) /* strip 9 char remote seq */))
            // TODO: allow for mixed modDates
            // use set subtraction remote filenames from local filenames
            for (const filename of filenames) {
                if (filename.endsWith('-lost.sltt-doc')) {
                    continue
                }
                const strippedFilename = filename.slice(9) /* strip 9 char local-doc */
                if (!strippedRemoteFilenames.has(strippedFilename)) {
                    localFilenames.push(filename)
                }
            }
            return localFilenames
        } else {
            return filenames
        }
    } catch (error) {
        console.error('An error occurred:', error.message)
        throw error
    }
}

export const handleRetrieveDocV0 = async (docsFolder: string, { clientId, project, isFromRemote, filename }: RetrieveDocArgs):
    Promise<RetrieveDocResponse<IDBObject> | null> => {
    const { normalizedFilename, remoteSeq, filenameModDate, filenameId, filenameCreator, filenameModBy } = parseFilename(filename)
    const fullFromPath = buildDocFolder(docsFolder, project, isFromRemote)
    const fullPath = join(fullFromPath, normalizedFilename)
    const { projectPath } = parseFilename(fullPath)
    try {
        const buffer = await readFile(fullPath)
        const doc = JSON.parse(buffer.toString())
        return { projectPath, remoteSeq, normalizedFilename, doc, fullPath, filenameId, filenameModDate, filenameCreator, filenameModBy }
    } catch (error) {
        if (error.code === 'ENOENT') {
            return null
        } else {
            console.error('An error occurred:', error.message)
            throw error
        }
    }
}

async function writeDoc(fullPath: string, doc: unknown):
    Promise<StoreDocResponse> {
    try {
        await writeFile(fullPath, JSON.stringify(doc))
        return { ...parseFilename(fullPath), freshlyWritten: true }
    } catch (error) {
        console.error('An error occurred:', error.message)
        throw error
    }
}

async function listDocs(docsFolder: string, { project, isFromRemote, fnFilter }: { project: string, isFromRemote: boolean, fnFilter?: (string) => boolean }): Promise<string[]> {
    const fullFromPath = buildDocFolder(docsFolder, project, isFromRemote)
    // detect if path doesn't yet exist
    if (!existsSync(fullFromPath)) {
        return []
    }
    console.log('listDocs fullFromPath:', fullFromPath)
    try {
        const filenames = await readdir(fullFromPath)
        console.log('filenames:', filenames)
        const result = filenames
            .filter(filename => filename.endsWith('.sltt-doc') && (!fnFilter || fnFilter(filename)))
        result.sort() // just in case it's not yet by name
        console.log('listDocs result:', result)
        return result
    } catch (error) {
        if (error.code === 'ENOENT') {
            return []
        } else {
            console.error('An error occurred:', error.message)
            throw error
        }
    }
}
