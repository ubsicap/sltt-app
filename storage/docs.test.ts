import { describe, it, expect, afterEach, beforeEach, vi } from 'vitest'
import { mkdtempSync, rmdirSync, existsSync } from 'fs'
import { join, resolve } from 'path'
import { tmpdir } from 'os'
import { handleListDocsV0, handleRetrieveDocV0, handleStoreDocV0, handleStoreRemoteDocs, IDBObject } from './docs'
import { mkdtemp, readFile, writeFile } from 'fs/promises'
import { ensureDir, remove } from 'fs-extra'
import { StoreRemoteDocsArgs, StoreRemoteDocsResponse } from './docs.d'

let tempDir: string

beforeEach(() => {
  // Create a unique temporary directory
  tempDir = mkdtempSync(join(tmpdir(), 'sltt-app-vitest-'))
})

afterEach(() => {
  // Clean up the temporary directory
  rmdirSync(tempDir, { recursive: true })
})

it('should create a temp folder path', () => {
  // Use the tempDir in your tests
  expect(existsSync(tempDir)).toBe(true)
})

describe('handleListDocs', () => {
  it('should list empty docs in empty folder', async () => {
    const project = 'testProject'
    const isFromRemote = false
    const testDataPath = resolve(__dirname, './test-data/listTests/empty')
    const docs = await handleListDocsV0(testDataPath, { clientId: 'client1', project, isFromRemote })
    expect(docs).toEqual([])
  })
  it('should strip remote docs, but list earlier local docs', async () => {
    const project = 'testProject'
    const isFromRemote = false
    const testDataPath = resolve(__dirname, './test-data/listTests/local-and-remote')
    const docs = await handleListDocsV0(testDataPath, { clientId: 'client1', project, isFromRemote })
    expect(docs).toEqual([
      'local-doc__2024-07-25_14-50-23-046__210629_180535-240725_145023__c62114c2__c62114c2.sltt-doc',
      'local-doc__2024-07-26_02-58-31-902__210629_180535-240726_025831__c62114c2__c62114c2.sltt-doc',
    ])
  })
  it('should list all remote docs', async () => {
    const project = 'testProject'
    const isFromRemote = true
    const testDataPath = resolve(__dirname, './test-data/listTests/local-and-remote')
    const docs = await handleListDocsV0(testDataPath, { clientId: 'client1', project, isFromRemote })
    expect(docs).toEqual([
      '000000001__2024-07-25_16-26-36-672__210202_183235-240607_145904-240618_160543-240725_162634__e85c7697__e85c7697.sltt-doc',
    ])
  })
})

describe('handleRetrieveDoc', () => {
  it('should retrieve documents correctly', async () => {
    const project = 'testProject'
    const filename = 'local-doc__2024-07-25_14-50-23-046__210629_180535-240725_145023__c62114c2__c62114c2.sltt-doc'
    const isFromRemote = false
    const testDataPath = resolve(__dirname, './test-data/listTests/local-and-remote')
    const response = await handleRetrieveDocV0(testDataPath, { clientId: 'client1', project, isFromRemote, filename })
    expect(response).toBeTruthy()
    const { fullPath } = response!
    const relativePath = fullPath.split('storage').pop() || ''
    const storagePath = join('storage', relativePath)
    const normalizedStoragePath = storagePath.replace(/\\/g, '/')
    expect({
      ...response,
      fullPath: normalizedStoragePath
    }).toMatchSnapshot()
  })

  it.each([
    {
      testCase: 'store remote doc with mod-by',
      project: 'testProject1',
      doc: {     
        modDate: '2024/06/21 06:05:21.444Z',
        _id: '210202_183235/240607_145904',
        creator: 'wendy@example.com',
        modBy: 'bob@example.com',
      },
      remoteSeq: '000000001',
      expectedFilename: '000000001__2024-06-21_06-05-21-444__210202_183235-240607_145904__3de71188__4b9bb806.sltt-doc'
    },
    {
      testCase: 'store local doc',
      project: 'testProject1',
      doc: {
        modDate: '2023/06/21 06:15:21.444Z',
        _id: '310202_183235/340607_145904',
        creator: 'wendy@example.com',
        modBy: 'bob@example.com',
      },
      remoteSeq: '',
      expectedFilename: 'local-doc__2023-06-21_06-15-21-444__310202_183235-340607_145904__3de71188__4b9bb806.sltt-doc'
    }
  ])('should return null for $testCase because files are not found', async ({ project, doc, remoteSeq, expectedFilename }) => {
    const isFromRemote = !remoteSeq
    const testDataPath = resolve(__dirname, './test-data/listTests/local-and-remote')
    const response = await handleRetrieveDocV0(testDataPath, { clientId: 'client1', project, isFromRemote, filename: expectedFilename })
    expect(response).toBe(null)
  })
})

describe('handleStoreDoc', () => {
  // Store 3 separate remote and local docs and verify they were saved
  it.each([
    {
      testCase: 'store remote doc with no-mod-by',
      project: 'testProject1',
      doc: {
        modDate: '2023/10/01 12:34:23.046Z',
        _id: '210202_183235/240607_145904',
        creator: 'bob@example.com',
      },
      remoteSeq: 1,
      expectedFilename: '000000001__2023-10-01_12-34-23-046__210202_183235-240607_145904__4b9bb806__no-mod-by.sltt-doc'
    },
    {
      testCase: 'store remote doc with modBy',
      project: 'testProject1',
      doc: {
        modDate: '2023/10/01 12:34:23.046Z',
        _id: '210202_183235/240607_145904',
        creator: 'alice@example.com',
        modBy: 'bob@example.com',
      },
      remoteSeq: 1,
      expectedFilename: '000000001__2023-10-01_12-34-23-046__210202_183235-240607_145904__c160f8cc__4b9bb806.sltt-doc'
    },
    {
      testCase: 'store local doc',
      project: 'testProject2',
      doc: {
        modDate: '2023/11/01 13:34:23.046Z',
        _id: '310202_183235/340607_145904',
        creator: 'alice@example.com',
        modBy: 'bob@example.com',
      },
      remoteSeq: Number.NaN,
      expectedFilename: 'local-doc__2023-11-01_13-34-23-046__310202_183235-340607_145904__c160f8cc__4b9bb806.sltt-doc'
    }
  ])('should handle document storage correctly for $testCase', async ({ project, doc, remoteSeq, expectedFilename }) => {
    const response = await handleStoreDocV0(tempDir, { clientId: 'client1', project, doc, remoteSeq })

    // Split the filename correctly
    const parts = expectedFilename.split('.')[0].split('__')
    const [expectedRemoteSeq, expectedFilenameModDate, expectedFilenameId, expectedFilenameCreator, expectedFilenameModBy] = parts
    const projectPath = `${project}/${!remoteSeq ? 'local' : 'remote'}`
    expect(response).toEqual({
      projectPath,
      normalizedFilename: expectedFilename,
      remoteSeq: expectedRemoteSeq,
      filenameModDate: expectedFilenameModDate,
      filenameId: expectedFilenameId,
      filenameCreator: expectedFilenameCreator,
      filenameModBy: expectedFilenameModBy,
      freshlyWritten: true
    })
  })

  it('should delete local doc after its remote is stored', async () => {
    const project = 'testProject1'
    const doc = {
      modDate: '2024/06/29 11:35:22.044Z',
      _id: '210202_183235/240607_145904',
      creator: 'wendy@example.com',
    }
    const expectedFilename1 = 'local-doc__2024-06-29_11-35-22-044__210202_183235-240607_145904__3de71188__no-mod-by.sltt-doc'
    const expectedFilename2 = '000000001__2024-06-29_11-35-22-044__210202_183235-240607_145904__3de71188__no-mod-by.sltt-doc'
    const firstStoreResponse = await handleStoreDocV0(tempDir, { clientId: 'client1', project, doc, remoteSeq: Number.NaN })
    const { projectPath, normalizedFilename } = firstStoreResponse
    const localPath = join(tempDir, projectPath, normalizedFilename)
    const localFileExists = existsSync(localPath)
    expect(localFileExists).toBe(true)
    expect(firstStoreResponse.freshlyWritten).toBe(true)
    expect(firstStoreResponse.normalizedFilename).toBe(expectedFilename1)

    const secondStoreResponse = await handleStoreDocV0(tempDir, { clientId: 'client1', project, doc, remoteSeq: 1 })
    const { projectPath: projectPath2, normalizedFilename: normalizedFilename2 } = secondStoreResponse
    const remotePath = join(tempDir, projectPath2, normalizedFilename2)
    const remoteFileExists = existsSync(remotePath)
    expect(remoteFileExists).toBe(true)
    expect(secondStoreResponse.freshlyWritten).toBe(true)
    expect(secondStoreResponse.normalizedFilename).toBe(expectedFilename2)
    const localFileExists2 = existsSync(localPath)
    expect(localFileExists2).toBe(false)
  })

  it('should not store the same doc twice', async () => {
    const project = 'testProject1'
    const doc = {
      modDate: '2024/07/30 12:34:23.046Z',
      _id: '210202_183235/240607_145904',
      creator: 'ellis@example.com',
    }
    const expectedFilename = 'local-doc__2024-07-30_12-34-23-046__210202_183235-240607_145904__8cf5a227__no-mod-by.sltt-doc'

    const firstStoreResponse = await handleStoreDocV0(tempDir, { clientId: 'client1', project, doc, remoteSeq: Number.NaN })
    expect(firstStoreResponse.freshlyWritten).toBe(true)
    expect(firstStoreResponse.normalizedFilename).toBe(expectedFilename)

    const secondStoreResponse = await handleStoreDocV0(tempDir, { clientId: 'client1', project, doc, remoteSeq: Number.NaN })
    expect(secondStoreResponse.freshlyWritten).toBe(false)
    expect(secondStoreResponse.normalizedFilename).toBe(expectedFilename)
  })
})

describe('handleStoreRemoteDocs', () => {
  let tempDir: string

  beforeEach(async () => {
    // Create a temporary directory for each test
    tempDir = await mkdtemp(join(tmpdir(), 'docs-'))
  })

  afterEach(async () => {
    // Clean up the temporary directory after each test
    await remove(tempDir)
  })
  // C:\\Users\\ericd\\AppData\\Local\\Temp\\docs-3SG87D\\testProject\\remote\\remote.sltt-docs

  it('should handle the empty seqDocs', async () => {
    const docsFolder = tempDir
    const clientId = 'testClient'
    const project = 'testProject'
    const seqDocs: StoreRemoteDocsArgs<IDBObject>['seqDocs'] = []

    const args: StoreRemoteDocsArgs<IDBObject> = { clientId, project, seqDocs }
    const response: StoreRemoteDocsResponse = await handleStoreRemoteDocs(docsFolder, args)

    // Check that the response contains no new lines
    expect(response).toEqual({ lastSeq: -1, storedCount: 0 })
  })

  it('should append new lines for incoming seqDocs with higher sequence numbers', async () => {
    const docsFolder = tempDir
    const clientId = 'tscl'
    const project = 'testProject'
    const seqDocs: StoreRemoteDocsArgs<IDBObject>['seqDocs'] = [
      { doc: { _id: '20240917', modDate: '2024/09/17 12:30:33', creator: 'bob@example.com' }, seq: 1 },
      { doc: { _id: '20240917', modDate: '2024/09/17 12:30:34', creator: 'bob@example.com' }, seq: 2 },
    ]

    const remoteSeqDocsFile = join(docsFolder, project, 'remote', 'remote.sltt-docs')

    const args: StoreRemoteDocsArgs<IDBObject> = { clientId, project, seqDocs }
    const response: StoreRemoteDocsResponse = await handleStoreRemoteDocs(docsFolder, args)

    // Check that the remote file was updated with the new lines
    const fileContent = await readFile(remoteSeqDocsFile, 'utf-8')
    const allLines = fileContent.split('\n')
    expect(allLines).toEqual([
      expect.stringMatching(/^000000001\t\d{13}\ttscl\t{"_id":"20240917","modDate":"2024\/09\/17 12:30:33","creator":"bob@example.com"}\t000000001$/),
      expect.stringMatching(/^000000002\t\d{13}\ttscl\t{"_id":"20240917","modDate":"2024\/09\/17 12:30:34","creator":"bob@example.com"}\t000000002$/),
      ''
    ])

    expect(response).toEqual({ lastSeq: 2, storedCount: 2 })
  })
})
