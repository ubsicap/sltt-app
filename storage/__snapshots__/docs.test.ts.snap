// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleRetrieveDoc > should retrieve documents correctly 1`] = `
{
  "doc": {
    "_id": "210629_180535/240725_145023",
    "assignee": "",
    "contentType": "Translation",
    "creationDate": "2024/07/25 14:50:23.046Z",
    "creator": "epyle@biblesocieties.org",
    "difficulty": 1,
    "modBy": "epyle@biblesocieties.org",
    "modDate": "2024/07/25 14:50:23.046Z",
    "name": "test offline passage",
    "rank": "00000001000.0000",
    "references": "[]",
  },
  "filename": "local-doc__2024-07-25_14-50-23-046__210629_180535-240725_145023__c62114c2__c62114c2.sltt-doc",
  "filenameCreator": "c62114c2",
  "filenameId": "210629_180535-240725_145023",
  "filenameModBy": "c62114c2",
  "filenameModDate": "2024-07-25_14-50-23-046",
  "fullPath": "storage\\test-data\\listTests\\local-and-remote\\testProject\\local\\local-doc__2024-07-25_14-50-23-046__210629_180535-240725_145023__c62114c2__c62114c2.sltt-doc",
  "remoteSeq": "local-doc",
}
`;
